server/server_test.go

```diff
@@ -29,38 +32,476 @@ func performRequest(r http.Handler, method, path string, body io.Reader) (*httpt
+			_, ok := game.PlayerColors[p.ID]
+			assert.True(t, ok)
+		}
 	}
-	cs := &cribbageServer{}
-	router := cs.NewRouter()
+}
+func TestGinPostCreateInteraction(t *testing.T) {
+	testCases := []struct {
+		msg     string
+		pIDs    []string
+		reqData network.CreateInteractionRequest
+		expCode int
+		expErr  string
+	}{{
+		msg: `missing player ID`,
+		reqData: network.CreateInteractionRequest{
+			PlayerID:      ``,
+			LocalhostPort: `1234`,
+		},
+		expCode: http.StatusBadRequest,
+		expErr:  `Needs playerId`,
+	}, {
+		msg:     `bad request body - equivalent to an empty network.CreateInteractionRequest`,
+		reqData: network.CreateInteractionRequest{},
+		expCode: http.StatusBadRequest,
+		expErr:  `Needs playerId`,
+	}, {
+		msg: `good request`,
+		reqData: network.CreateInteractionRequest{
+			PlayerID:      `p1`,
+			LocalhostPort: `1234`,
+		},
+		expCode: http.StatusOK,
+		expErr:  ``,
+	}, {
+		msg: `unsupported interaction mode`,
+		reqData: network.CreateInteractionRequest{
+			PlayerID: `p1`,
+		},
+		expCode: http.StatusBadRequest,
+		expErr:  `unsupported interaction mode`,
+	}}
+	cs, router := newServerAndRouter()
+	seedPlayers(t, cs.db, 5)
 	for _, tc := range testCases {
-		reqBytes, err := json.Marshal(tc.data)
+		// make the request
+		body := prepareBody(t, tc.reqData)
+		w, err := performRequest(router, `POST`, `/create/interaction`, body)
 		require.NoError(t, err)
-		body := bytes.NewReader(reqBytes)
-		w, err := performRequest(router, `POST`, `/create/player`, body)
+		// verify
+		require.Equal(t, tc.expCode, w.Code)
+		if tc.expCode != http.StatusOK {
+			errMsg := readError(t, w)
+			assert.Equal(t, tc.expErr, errMsg)
+			continue
+		}
+		bs, err := ioutil.ReadAll(w.Body)
 		require.NoError(t, err)
-		assert.Equal(t, tc.expCode, w.Code)
-		if tc.expCode == http.StatusOK {
-			var player model.Player
-			readBody(t, w.Body, &player)
-			assert.NoError(t, err)
-			assert.Equal(t, tc.expRes, player)
+		msg := string(bs)
+		// verify the players are in the game
+		assert.Equal(t, `Updated player interaction`, msg)
+	}
+}
+func TestGinGetGame(t *testing.T) {
+	testCases := []struct {
+		msg     string
+		gameID  string
+		expCode int
+		expErr  string
+	}{{
+		msg:     `bad game ID`,
+		gameID:  `123zzz`,
+		expCode: http.StatusBadRequest,
+		expErr:  `Invalid GameID: strconv.Atoi: parsing "123zzz": invalid syntax`,
+	}, {
+		msg:     `normal request`,
+		gameID:  ``,
+		expCode: http.StatusOK,
+		expErr:  ``,
+	}, {
+		msg:     `nonexistent game`,
+		gameID:  `123`,
+		expCode: http.StatusNotFound,
+		expErr:  `Game not found`,
+	}}
+	cs, router := newServerAndRouter()
+	pIDs := seedPlayers(t, cs.db, 2)
+	for _, tc := range testCases {
+		// seed the db with a game
+		g, err := createGame(context.Background(), cs.db, pIDs)
+		require.NoError(t, err)
+		// make the request
+		var url string
+		if tc.gameID != `` {
+			url = `/game/` + tc.gameID
+		} else {
+			url = `/game/` + fmt.Sprintf(`%d`, g.ID)
+		}
```

Comment:
Instead, could we provide a `setup` func on the test case which we call and then potentially returns the url for us?

```go
testCases := []struct {
  ...
  setup func(db, pIDs) (url string, g model.Game)
}{{ ...
```

Reply:
I feel like this will result in more copy/pasted code since each `setup` will have
```go
// seed the db with a game
g, err := createGame(context.Background(), cs.db, pIDs)
require.NoError(t, err)
```
But I could be persuaded - what's the advantage?

Reply:
🤷 I can go either way. doing a setup func means you don't need to have conditions in the "setup portion" vs. sometimes it's harder to wrap your mind around. 🤷

Reply:
I am persuaded